web/controllers/api/docs_controller.ex

```diff
@@ -163,16 +163,13 @@ defmodule HexWeb.API.DocsController do
     HexWeb.CDN.purge_key(:fastly_hexdocs, unversioned_key)
     HexWeb.CDN.purge_key(:fastly_hexdocs, versioned_key)
 
-    # Set docs flag on release
-    HexWeb.Repo.transaction(fn ->
-      Ecto.Changeset.change(release, has_docs: true)
-      |> HexWeb.Repo.update!
+    multi =
+      Ecto.Multi.new
+      |> Ecto.Multi.update(:release, Ecto.Changeset.change(release, has_docs: true))
+      |> Ecto.Multi.update(:package, Ecto.Changeset.change(release.package, docs_updated_at: Ecto.DateTime.utc))
+      |> Ecto.Multi.run(:log, fn _ -> audit(user, "docs.publish", {package, release}) end)
```

Comment:
> logger_id would be a virtual field. And yes, there would be a changeset function that would be invoked by the put_assoc with AuditLog.

I'm probably missing something, but it looks like `AuditLog.changeset` isn't invoked by doing `release_changeset.put_assoc(:audit_log)`. It would have been invoked when doing `release_changeset.cast_assoc(:audit_log)` though. I'm not sure if `cast_assoc` would work well here, as we would need to put auditing log data into params that come into the changeset.

I'm still not 100% sure how the virtual `logger_id` would have been automatically set - would Ecto automatically do that based on association informations, or is there some hook that can be used?

> Maybe we should allow setting defaults to associations. I can see many places where starting with them as nil or an empty list would be useful. Thoughts?

I think this would be helpful in this particular case, but I'm not sure about other use cases - it would be great to enumerate them first. Btw, would you allow setting just `nil` and `[]` or actual data as well, as in arbitrary records or arrays of records?

Thanks for your patience!